File: back-end/src/main/java/org/scutsu/materialmanagement/backend/models/Person.java
package org.scutsu.materialmanagement.backend.models;

import lombok.Data;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
public class Person {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    @NotNull
    private String name;

    @ManyToOne
    private Organization organization;

    @ElementCollection
    private List<PersonMetadata> metadata = new ArrayList<>();
}

@Data
@Embeddable
class PersonMetadata {

    @ManyToOne
    private PersonMetadataType type;

    private String value;
}

File: back-end/src/main/kotlin/org/scutsu/materialmanagement/backend/models/Person.kt
package org.scutsu.materialmanagement.backend.models

import java.io.Serializable
import javax.persistence.*

@Entity
data class Person(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long,

    @Column(nullable = false)
    val name: String,

    @ManyToOne
    val organization: Organization,

    @ElementCollection
    val metadata: List<PersonMetadata> = ArrayList()
)

@Entity
data class PersonMetadataType(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long,

    @Column(nullable = false)
    val name: String
)

@Embeddable
data class PersonMetadata(
    @ManyToOne
    val type: PersonMetadataType,

    val value: String
): Serializable
